Python/Product/PythonTools/PythonTools/Environments/AddCondaEnvironmentOperation.cs

```diff
@@ -0,0 +1,281 @@
+            var tcs = new TaskCompletionSource<IPythonInterpreterFactory>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            EventHandler interpretersChanged = (object sender, EventArgs e) => {
+                var first = _registry.Interpreters
+                    .Where(f => CondaEnvironmentFactoryProvider.IsCondaEnv(f, _envName))
+                    .FirstOrDefault();
+                if (first != null) {
+                    tcs.TrySetResult(first);
+                }
+            };
+
+            IPythonInterpreterFactory factory = null;
+
+            _registry.InterpretersChanged += interpretersChanged;
+            try {
+                taskHandler.Progress.Report(new TaskProgressData() {
+                    CanBeCanceled = false,
+                    ProgressText = Strings.CondaStatusCenterCreateProgressCreating,
+                    PercentComplete = null,
+                });
+
+                bool failed = false;
+                bool useEnvFile = !string.IsNullOrEmpty(_envFilePath);
+
+                try {
+                    if (useEnvFile) {
+                        if (!await _condaMgr.CreateFromEnvironmentFileAsync(
+                            _envName,
+                            _envFilePath,
+                            ui,
+                            ct
+                        )) {
+                            throw new ApplicationException(Strings.CondaStatusCenterCreateFailure);
+                        }
+                    } else {
+                        if (!await _condaMgr.CreateAsync(
+                            _envName,
+                            _packages.ToArray(),
+                            ui,
+                            ct
+                        )) {
+                            throw new ApplicationException(Strings.CondaStatusCenterCreateFailure);
+                        }
+                    }
+                } catch (Exception ex) when (!ex.IsCriticalException()) {
+                    failed = true;
+                    throw;
+                } finally {
+                    _logger?.LogEvent(PythonLogEvent.CreateCondaEnv, new CreateCondaEnvInfo() {
+                        Failed = failed,
+                        FromEnvironmentFile = useEnvFile,
+                        SetAsDefault = _setAsDefault,
+                        SetAsCurrent = _setAsCurrent,
+                        OpenEnvironmentsWindow = _viewInEnvWindow,
+                    });
+                }
+
+                // Wait for the factory to refresh
+                taskHandler.Progress.Report(new TaskProgressData() {
+                    CanBeCanceled = false,
+                    ProgressText = Strings.CondaStatusCenterCreateProgressDetecting,
+                    PercentComplete = 90,
+                });
+
+                var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
+                if (await Task.WhenAny(tcs.Task, timeoutTask) == timeoutTask) {
```

Comment:
You just need to remember that it isn't 100% universal. If continued task can actually be canceled, continuation task will be faulted instead of canceled because of this: https://github.com/dotnet/coreclr/blob/913428d5915a9729c9405f57a75e7f912f9d29a5/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2505 (.net framework implementation has similar code).

This is not a problem for a code like `await task.ContinueWith(t => t.GetAwaiter().GetResult(), ct)`, because the calling async code will convert it back into cancellation, but if you want to have an extension method, you can either use `Unwrap`:
```cs
task.ContinueWith(t => t, ct, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default).Unwrap();
```
which preserves the state, but requires one more instance of `Task`, or use `TaskContinuationOptions.NotOnCanceled`:
```
task.ContinueWith(t => t.GetAwaiter().GetResult(), ct, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.NotOnCanceled, TaskScheduler.Default);
```
which has a problem of swallowing `CancellationToken` from original `task`, but there are other cases when it is swallowed.

`TaskContinuationOptions.ExecuteSynchronously` is kinda desirable, cause it saves unnecessary scheduling, especially when `task` is already completed.